pkg/sql/importer/import_processor.go

```diff
@@ -517,22 +393,11 @@ func ingestKvs(
 		prog.ResumePos = make(map[int32]int64)
 		prog.CompletedFraction = make(map[int32]float32)
 		for file, offset := range offsets {
-			pk := atomic.LoadInt64(&pkFlushedRow[offset])
-			idx := atomic.LoadInt64(&idxFlushedRow[offset])
-			// On resume we'll be able to skip up the last row for which both the
-			// PK and index adders have flushed KVs.
-			if idx > pk {
```

Comment:
There is an interesting implication of this implementation where we have separate checkpoints for indexes/primary keys: the rest of the import pipeline needs to be tolerant of duplicate kv's since we might end up flushing a second copy of an index/primary key row on retry.

Reply:
The relevant behavior is on the add sst command in the KV layer. Import sets DisallowShadowingBelow to the time the import starts, so it will ignore duplicate keys if the row was written by import and has the same value.
 https://github.com/cockroachdb/cockroach/blob/fa5d244500f8c67827d3ff5694db882fe0f54185/pkg/kv/kvpb/api.proto#L2093